demos/adjoint/adjoint_demo.py

```diff
@@ -0,0 +1,348 @@
+
+reduced_functional = ReducedFunctional(objective, control)
+
+# Pausing Annotation
+# ----------------------
+#
+# At this point, we have completed annotating the tape with the necessary information from running the forward simulation. To prevent further annotations during subsequent operations, we stop the annotation process. This ensures that no additional solves are unnecessarily recorded, keeping the tape focused only on the essential steps.
+
+pause_annotation()
+
+# We can then print the contents of the tape to verify that it is not empty.
+
+# + tags=["active-ipynb"]
+# print(tape.get_blocks())
+# -
+
+# Verification of Gradients: Taylor Remainder Convergence Test
+# ----------------------
+#
+# A fundamental tool for verifying gradients is the Taylor remainder convergence test. This test helps ensure that the gradients computed by our optimisation algorithm are accurate. For the reduced functional, $J(T_{ic})$, and its derivative, $\frac{\mathrm{d} J}{\mathrm{d} T_{ic}}$, the Taylor remainder convergence test can be expressed as:
+#
+# $$ \left| J(T_{ic} + h \,\delta T_{ic}) - J(T_{ic}) - h\,\frac{\mathrm{d} J}{\mathrm{d} T_{ic}} \cdot \delta T_{ic} \right| \longrightarrow 0 \text{ at } O(h^2). $$
+#
+# The expression on the left-hand side is termed the second-order Taylor remainder. This term's convergence rate of $O(h^2)$ is a robust indicator for verifying the computational implementation of the gradient calculation. Essentially, if you halve the value of $h$, the magnitude of the second-order Taylor remainder should decrease by a factor of 4.
+#
+# We employ these so-called *Taylor tests* to confirm the accuracy of the determined gradients. The theoretical convergence rate is $O(2.0)$, and achieving this rate indicates that the gradient information is accurate down to floating-point precision.
+#
+# ### Performing Taylor Tests
+#
+# In our implementation, we perform a second-order Taylor remainder test for each term of the objective functional. The test involves computing the functional and the associated gradient when randomly perturbing the initial temperature field, $T_{ic}$, and subsequently halving the perturbations at each level.
+#
+# Here is how you can perform a Taylor test in the code:
+
+# + tags=["active-ipynb"]
+# # Define the perturbation in the initial temperature field
+# import numpy as np
+# Delta_temp = Function(Tic.function_space(), name="Delta_Temperature")
+# Delta_temp.dat.data[:] = np.random.random(Delta_temp.dat.data.shape)
+#
+# # Perform the Taylor test to verify the gradients
+# minconv = taylor_test(reduced_functional, Tic, Delta_temp)
+# -
+
+# The `taylor_test` function computes the Taylor remainder and verifies that the convergence rate is close to the theoretical value of $O(2.0)$. This ensures that our gradients are accurate and reliable for optimisation.
```

Comment:
Is it worth providing the user with some guidance here on what this test spits out - and maybe even a quick plot of the the remainders or something, just to make it more interactive?

Reply:
Having now run it, I think it's worth making it clear to the user that this Taylor test takes time. It took over 20 minutes for me, so important they are aware of this so that they don't feel that they have done something wrong. Also I think flagging what the output at the bottom means would be good (computed residuals, computed convergence). Yes it's obvious to us - but it may not be to somebody coming at this for the first time.

Reply:
The issue here is that we cannot easily retrieve the residuals + perturbations from taylor test to automatically plot them, it all gets dumped to stdout. Unless I redirect `sys.stdout` to a variable, which I think would make the whole thing confusing!